src/proxy/http/mod.rs

```diff
@@ -62,14 +63,32 @@ impl HasH2Reason for ::h2::Error {
 pub fn authority_from_header<B, K>(req: &http::Request<B>, header: K) -> Option<Authority>
 where
     K: AsHeaderName,
+{
+    header_value_from_request(req, header, |s: &str| s.parse::<Authority>().ok())
+}
+
+pub fn identity_from_header<B, K>(req: &http::Request<B>, header: K) -> Option<identity::Name>
+where
+    K: AsHeaderName,
+{
+    header_value_from_request(req, header, |s: &str| {
+        identity::Name::from_hostname(s.as_bytes()).ok()
+    })
+}
+
+fn header_value_from_request<B, K, F, T>(
+    req: &http::Request<B>,
+    header: K,
+    try_from: F,
+) -> Option<T>
+where
+    K: AsHeaderName,
+    F: FnOnce(&str) -> Option<T>,
 {
     req.headers().get(header).and_then(|value| {
-        value.to_str().ok().and_then(|s| {
-            if s.is_empty() {
-                None
-            } else {
-                s.parse::<Authority>().ok()
-            }
-        })
+        value
+            .to_str()
+            .ok()
+            .and_then(|s| if s.is_empty() { None } else { try_from(s) })
```

Comment:
I'm not sure that this helper buys all that much, but if we preserve it it would be clearer using the short-circuit syntax:
```rust
let value = req.headers().get(header)?.to_str().ok()?;
if value.is_empty()  {
    return None;
}
try_from(s)
```

Reply:
I can change the syntax. I added it because when I added the other `fn`, I had both right next to each other and they did pretty much the exact same thing.